File: open_closed.ts
// Open Closed Principle
// it means that the class is open to extends and close to change

//////////////// 
//  WROOOONG   / 
//////////////// 
class Superhero_ {
    fly(): string {
        return "fly"
    }
    superStrength(): string {
        return "super strength"
    }

    // xRayVision // it is broken open closed principle
    // superJump
    // invisbility
}

// const superman = new Superhero_()
// superman.fly()
// superman.superStrength()

// const hulk = new Superhero_()
// hulk.fly()

//////////////// 
//  BETTER     / 
//////////////// 
abstract class SuperSkill {
    abstract do(): string
}

class Fly implements SuperSkill {
    do(): string {
        return 'fly'
    }
}

class SuperStrength implements SuperSkill {
    do(): string {
        return 'super strength'
    }
}

class Invisbility implements SuperSkill {
    do(): string {
        return 'invisbility'
    }
}

class Superhero {
    superSkills: SuperSkill[]

    constructor(superSkills: SuperSkill[]) {
        this.superSkills = superSkills
    }
}

const fly = new Fly()
const superst = new SuperStrength()

const superman = new Superhero([fly, superst])

const hulk = new Superhero([superst])

console.log('superman => ', superman.superSkills)
console.log('hulk => ', hulk.superSkills)

File: python/open_closed.py
# Open Closed Principle
# it means that the class is open to extends and close to change

"""
WRONG
"""
class SuperHeroWrong:
    def __init__(self, name):
        self.name = name

    def fly(self) -> str:
        return f'{self.name} -> flying'
    
    def super_strength(self) -> str:
        return f'{self.name} -> super strength'
    

"""
BETTER
"""
# class SuperHero:
# create a new version without breaking OCP

# Clase de poderes
class PowerManager:
    def __init__(self, skills_array):
        self.skills = skills_array

    def execute(self):
        if not self.skills:
            return "no tiene habilidades asignadas."
        return "usando " + " y ".join(self.skills)

# La clase Padre (SuperHero)
class SuperHero:
    def __init__(self, name, skills_list):
        self.name = name
        self.super_poder = PowerManager(skills_list)

    def activar_superPoder(self):
        return f"{self.name} está {self.super_poder.execute()}"
        
# Hago que el nombre y las habilidades del SuperHero las elija el usuario que instancia, cosa de que si quiere hacer un 
# SuperHero nuevo llamado SuperMono que puede programar, lo pueda hacer.


# running process
superman = SuperHeroWrong('superman')
print(superman.fly())
print(superman.super_strength())

hulk = SuperHeroWrong('hulk')
print(hulk.super_strength())
### ???? fix it
print(hulk.fly())

 # what happen if I need to and invisible woman? how to add invisibility just for her?
